README
OUAC# 2012-156363

Versions of the bot starting from NumDiff requires the NumPy library to be installed in order to run.
Code only tested in Linux.

To simulate a game:
1. Edit ./play_match.sh options if needed (e.g. change "--turns" to larger number)
2. Choose one of following:
$ ./play_match.sh CombatBot.py
$ ./play_match.sh CombatBot.py --log=DEBUG  # log located at game_logs/CombatBot.log
$ ./v_play_match.sh CombatBot.py # launches debugging visualizer to display info for specific turns

To profile code (after simulating a game):
$ ./run_profiler.sh CombatBot.py
$ python view_profile.py

-----------------------------------------------------------

My work on the Fall 2011 AI Challenge, Ants, uses the default Python starter package as a base to work from. I used ideas from the Forums such as collaborative diffusion and Memetix's combat logic. I also used the AI-state-visualizer provided at https://github.com/j-h-a/aichallenge/blob/vis_overlay/VIS_OVERLAY.md

Username: VK_eipi
Rank: 850/7897
Description:
"The AI Challenge of 2011 was in many ways a similar experience to the one of 2010. However, due to the larger map sizes and greater number of decisions to made with the different game, I was confronted with a major weakness of Python: its speed. I constantly ran into the time limits, which forced me optimize my code and to look for creative solutions. For instance, I used the external library NumPy for faster numerical processing and implemented a “collaborative diffusion” map to guide my bot’s actions instead of extensive pathfinding. I continued to use technical code maintenance skills such as debugging, logging changes, and commenting."

Bot versions (detailed in bot docstrings and changelog.txt):
StarterBot.py - same behaviour as default starter package but code edited/refactored
TutorialBot.py - added behaviour of online tutorial but with better exploration code
AstarBot.py - added A* searches; too slow
DiffusionBot.py - switched to diffusion map for pathfinding
    HackDiffBot.py - quick changes to make performance submission-ready
NumDiffBot.py - began using NumPy library for optimization
    NumDiffBot_006.py - final official submission for AI Challenge
ExploreBot.py - increased exploration and "spreading out"
CombatBot.py - added combat logic, but introduced some flaws/bugs
             - contains most complete history in docstring


changelog.txt
NOTES
- 60x90 map -> 93-101 ms for Location objects
- 150x150 map -> 515ms for Location objects
- on 60x90 map, usually costs >3 seconds to generate ranges (timeout)
  - unknown reason for sometimes costing 2.2 seconds to generate ranges
    - in one case, 0.57ms/cell in part1 but 0.40ms/cell in part2
  - 0.53-0.57 milliseconds per cell (measured with half of rows)
  - 0.48-0.51 milliseconds per cell if only vision (half rows)
  - 0.50-0.53 milliseconds per cell if vision+gather (half rows)
- TutorialBot currently uses 1-5 ms/ant at beginning for exploration
  - for 60x90, peaks ~ 25ms then goes down once mostly explored

- if efficiency ever required, consider bitmask/field
  - try using numpy; use zip in Downloads as example
    - diffusion setup 7-10 times faster! (~70ms -> ~10ms on 60x90)
    - diffusion ~85 times faster! (~18.5-19ms -> 0.2-0.4ms per iteration on 60x90)


- suburb exploration decreases timeouts and improves obstacle navigation
  but the latter causes the bot to fail on tutorial map with improved
  ability to suicide charge

- scent maps: red to yellow/violet works well

- some maps on TCP, like cell_maze_p03_03.map, are too big by new standards
  - use alarm to timeout for etiquette
  - ExploreBot might not handle even close to max like 130x176!

TODO
- make sure visibility code is accurate
  - is version for smaller radius better than rolling?

- exploration scent: experiment with higher e_factor


- enemy ants emit scent; stronger as closer to hill (weaker if only one hill)
- defence: single-emit hill scent -> weak zones = low concentration
  - own ants block; enemies emit negative
- combat: Memetix style?

- possible inefficiencies:
  - Location.passable()
  - roll (esp. from do_turn like diffusion)
  - gen_vision_field
  - 
  

- check if vision field generation takes too much time with many ants
- consider using spare time to generate more ranges
  - store done/undone in a set?
- would arrays grant much greater efficiency?
  
- add testing : ie check vision, etc. actually works
- try https://github.com/j-h-a/aichallenge/blob/vis_overlay/VIS_OVERLAY.md
- Default Game for testing (200x200 is max size, maps/multi_hill_maze/maze_05p_01.map is 150x150)
  - 1hillx4p, 60x90: maze/maze_04p_01.map
  - 6hillsx3p, 108x144: multi_hill_maze/maze_03p_01.map
  - 5hillsx3p, 145x145:  maze/maze_p03_03.map
  - 1hillx4p, 124x120: random_walk/random_walk_p04_02.map

- MyAnt object with pos = Location
- assassin pairs

DONE

proj/numAnts.py
v2.1 (002)
- track invisible food
v2 (001)
- remove ranges in favor of numpy arrays and cell.adj

proj/numLocation.py
v2
- remove ranges
- stamp functions
- adj property

HackDiffBot
v1 (000)
- diffusion limit by time
- increase enemy hill scent
- reduce scent every turn
- diffuse function

ExploreBot
* v3, 3.1 (002)
- exploration diffusion
  - exploration scent reset every turn
  - unexplored edge is clamp; invisible just initialized
- food scent decreases over time
- ants emit repelling food scent
- tune factors / iterations
NumDiffBot
* v2.2 (006)
- use numAnts 
  - food is persistent
* v2.1
- clamps replace scents
- different visualization shading
* v2 (004)
- reset blocked scents
- 50 iterations
* v1.5
- use 2d arrays instead of iteration
DiffusionBot
* v1
- 10 diffusion iterations
- figured out color map
* v0.1 (002)
- implemented diffusion (unified)
- added custom visualization to options

AstarBot
* v1 (003)
- upper limit on A* searches
* v0 (002)
- A* pathfinding for food
- created Node class
 
ants.py
* v1.1.1 (015)
- gen_ranges() for newly seen land
- log_time function reporting used, left

TutorialBot
* v1.2 (010)
- regionalize function -> (sanctum, suburb)
- exploration uses suburb of explored with dynamic time limit
* v1.0 (005)
- Bot instance attributes: world, unknown sets
- Follows online tutorial:
  - collision detection
  - food gathering
  - evac hill
  - exploration
  - attack hills
 
StarterBot.py
- logging improvements

ants.py / location.py : Version 1.0 (ants_012, location_007)

CombatBot.py
#!/usr/bin/env python

"""
CombatBot
v4.3.2 (017)
- fixed bug: a_scent inconsistency due to adding to empty value
v4.3.1 (016)
- fixed bug: realized "stay in place" move will never be unoccupied_next
- fixed bug: changed self to ant_loc in movement code
- fixed bug when "stay in place" is NEUTRAL and best move
v4.3 (015)
- Memetix-style combat safety
- u_scent does not repel if soldiers or enemies
- clamp nearby enemy a_scents
- add enemies to exploration border
- decrease diffusion iterations to prevent timeout
- sort ants before movement 

v4.2
- hilldist (limit=15) and threat_value based on hilldist 
- change colour map code
v4.1 (006)
- Attack diffusion against enemy hills + ants with co-op factor
- eliminate unnecessary unoccupied_next setup
- iterations: U120, E120, A40
v4.0
- use concatenate & indexing instead of roll for speed boost
- rearrange unblocked /diff setup

ExploreBot
v3.2
- use passable/ant fields from numAnts v2.3
v3, v3.1 (002)
- exploration diffusion
  - exploration scent reset every turn
  - unexplored edge is clamp; invisible just initialized
- food scent decreases over time
- ants emit repelling food scent
- tune factors / iterations

NumDiffBot
v 2.2 (006)
- use numAnts 
  - food is persistent
v 2.1
- clamps replace scents
- different visualization shading
v2 (004)
- reset blocked scents
- 50 iterations
v1.5 (003)
- use 2d arrays instead of iteration

DiffusionBot
v1 (003)
- 10 diffusion iterations
- figured out color map
v0.1 (002)
- implemented diffusion (unified)
- added custom visualization to options
"""

import sys
import os
import time
import logging
from optparse import OptionParser
import copy
from math import sqrt

import numpy as np

from proj.numAnts import Ants
from proj.numLocation import or_stamp, increment_stamp, min_stamp
from proj.constants import LAND, WATER, UNKNOWN, ME, FOOD, INF, SAFE, NEUTRAL, DIE

class Settings:
    VISUALIZE = False
    LOGGING = False
log = logging.getLogger(__name__)

def init_options():
    parser = OptionParser()
    parser.add_option("-l", "--loglevel", "--log", dest="level", 
        type="choice", choices=("DEBUG", "INFO", "WARNING", "ERROR"),
        help="LEVEL defines minimum importance for logging. "
             "If not defined, no logging is done." )
    parser.add_option("-v", "--visual", action="store_true", 
                      dest="visualize", default=False, 
                      help="Turns on custom visualization.")
                     
    (options, args) = parser.parse_args()
    
    # setup visualizer
    Settings.VISUALIZE = options.visualize

    # setup logging
    level = {"DEBUG": logging.DEBUG,
             "INFO": logging.INFO,
             "WARNING": logging.WARNING,
             "ERROR": logging.ERROR} #logging.FATAL left out on purpose
    if options.level is not None:
        Settings.LOGGING = True
        LOGLEVEL = level[options.level]
        FILENAME = os.path.splitext(sys.argv[0].strip())[0]
        LOG_FILENAME = os.path.join("game_logs", FILENAME + '.log')
        if Settings.LOGGING:
            logging.basicConfig(filename=LOG_FILENAME,level=LOGLEVEL)
            
    log.info("\n\noptions: %s", options)

##def regionalize(lset):
    ##land = set(loc for loc in lset if loc.terrain is LAND)
    ##sanctum = set(loc for loc in land if loc.vision_range.issubset(lset))
    ##suburb = land - sanctum
    ##return (sanctum, suburb)

def roll_all(matrix):
    n = np.vstack((matrix[-1], matrix[:-1]))
    s = np.vstack((matrix[1:], matrix[0]))
    w = np.hstack((matrix[:,-1:], matrix[:,:-1]))
    e = np.hstack((matrix[:,1:], matrix[:,:1]))
    return (n, s, w, e)

class MyBot(object):
    def __init__(self):
        MyBot.DIFF_FACTOR = 0.2
        MyBot.U_ITERATIONS = 60
        MyBot.U_DECAY = 0.95
        
        MyBot.E_FACTOR = 0.15
        MyBot.E_ITERATIONS = 90
        
        MyBot.A_FACTOR = 0.2
        MyBot.COOP_FACTOR = 1.01
        MyBot.A_ITERATIONS = 40
        MyBot.A_DECAY = 0.9
    
    # do_setup is run once at the start of the game
    # after the bot has received the game settings
    def do_setup(self, ants):
        self.ants = ants
        ##self.world = set(c for r in ants.loc for c in r)
        
        if Settings.VISUALIZE:
            self.timer_cutoff = False
        
        self.u_scent_field = np.zeros(ants.dimensions) # unified
        self.e_scent_field = np.zeros(ants.dimensions) # exploration
        self.a_scent_field = np.zeros(ants.dimensions) # attack
        self.enemy_seen = False
        
        self.player_dimensions = (10, ants.rows, ants.cols)
        self.attack_field = np.empty(self.player_dimensions, dtype=int)
        self.weakness_field = np.empty(self.player_dimensions, dtype=int)
        self.lowest_enemy_weakness = np.empty(ants.dimensions, dtype=int)
        self.safety_field = np.empty(ants.dimensions, dtype=np.int8)
        self.soldiers = None
        self.normal_ants = None
        
        # calculate rough combat influence after 1 move
        radius = int(sqrt(ants.attackradius2)) + 1 # 1 longer
        approx_side = 2 * radius + 1
        approx_stamp = np.zeros((approx_side, approx_side), dtype=bool)
        or_stamp(approx_stamp, ants.attack_stamp, (radius, radius))
        or_stamp(approx_stamp, ants.attack_stamp, (radius, radius+1))
        or_stamp(approx_stamp, ants.attack_stamp, (radius, radius-1))
        or_stamp(approx_stamp, ants.attack_stamp, (radius+1, radius))
        or_stamp(approx_stamp, ants.attack_stamp, (radius-1, radius))
        self.approx_combat_stamp = approx_stamp
        
        self.worst_time_used = 0.0 
        log.debug("%s ms left after bot setup", ants.setup_time_remaining())
        
    # do turn is run once per turn
    def do_turn(self, ants):
        log.info("= TURN {0} - do_turn - BEGINS =".format(ants.cur_turn))
        
        self.soldiers = set()
        self.diffuse_attack(ants)
        self.normal_ants = ants.my_ants() - self.soldiers
        
        self.diffuse_food(ants)
        self.diffuse_explore(ants)
        
        scent_field = (self.u_scent_field + 
                       5.0 * self.e_scent_field + 
                       self.a_scent_field)
        safety = self.safety_field
        
        # Movement
        # - combat logic
        for ant_loc in sorted(self.soldiers):
            if ant_loc.adj:
                scent_field[ant_loc] = (sum(scent_field[a] 
                                           for a in ant_loc.adj) / 
                                        len(ant_loc.adj))
                moves = [move for move in ant_loc.adj
                                  if (move.unoccupied_next)] + [ant_loc]
                moves.sort(reverse=True,
                       key=lambda a: (scent_field[a], a))           
                neutral_moves = []
                for target in moves:
                    if safety[target] == SAFE:
                        if target != ant_loc:
                            direction = ant_loc.direction(target)[0]
                            ants.issue_order((ant_loc, direction))
                            target.unoccupied_next = False
                            ant_loc.unoccupied_next = True
                            #log.debug("%r to %r", ant_loc, target)
                        break
                    elif safety[target] == NEUTRAL:
                        if target.hilldist <= 8:
                            if target != ant_loc:
                                direction = ant_loc.direction(target)[0]
                                ants.issue_order((ant_loc, direction))
                                target.unoccupied_next = False
                                ant_loc.unoccupied_next = True
                            break
                        else:
                            neutral_moves.append(target)
                else: # no currently possible move is SAFE / hillsac
                    if neutral_moves:
                        target = neutral_moves[0]
                        if target != ant_loc:
                            direction = ant_loc.direction(target)[0]
                            ants.issue_order((ant_loc, direction))
                            target.unoccupied_next = False
                            ant_loc.unoccupied_next = True
                        #log.debug("%r to %r", ant_loc, target)                        
                    else:
                        self.normal_ants.add(ant_loc)
        
        # - regular movement
        for ant_loc in sorted(self.normal_ants):
            for target in sorted(ant_loc.adj, reverse=True,
                                key=lambda a: (scent_field[a], a)):
                if target.unoccupied_next:
                    direction = ant_loc.direction(target)[0]
                    ants.issue_order((ant_loc, direction))
                    target.unoccupied_next = False
                    ant_loc.unoccupied_next = True
                    #log.debug("%r to %r", ant_loc, target)
                    break
        
        ants.log_time("BEFORE VISUALIZE")
        log.info("VISUALIZE: %s", Settings.VISUALIZE)
        if Settings.VISUALIZE and 100  <= ants.cur_turn < 130:
            intensities = [[] for i in xrange(256)]
            for cell in ants.explored:
                txts = [ 
                         "u_scent: {0}".format(self.u_scent_field[cell]),
                         "e_scent: {0}".format(5.0*self.e_scent_field[cell]),
                         "a_scent: {0}".format(self.a_scent_field[cell]),
                         "scent: {0}".format(scent_field[cell]),
                         ##"weakness: {0}".format(self.weakness_field[0][cell]),
                         ##"best_enemy: {0}".format(self.lowest_enemy_weakness[cell]),
                         ##"safety_field: {0}".format(safety[cell]),
                        ]
                txt = "; ".join(txts)
                cmd = "i {0} {1} {2}\n".format(cell.r, cell.c, txt)
                sys.stdout.write(cmd)
                
                ##intensity = max(0,min(255, int((scent_field[cell]+5)/100*256)))            
                ##intensity = max(0,min(255, int((self.a_scent_field[cell])/50*256)))
                ##intensity = max(0,min(255, int((safety[cell]+1)/3*256)))
                ##intensities[intensity].append(cell)
            for i in xrange(256):
                if intensities[i]:
                    color = (255-i, 0, i, 0.5)
                    color = map(str, color)
                    cmd1 = "v setFillColor {0}\n".format(" ".join(color))
                    sys.stdout.write(cmd1)
                    for cell in intensities[i]:
                        cmd2 = "v tile {0.r} {0.c}\n".format(cell)
                        sys.stdout.write(cmd2)
            sys.stdout.flush()
        
        self.worst_time_used = max(self.worst_time_used, 
                        1000 * (time.time() - ants.turn_start_time) )
        ##log.debug("self.orders (to: from): %s", self.orders)
        log.info("Most used: %s ms", self.worst_time_used)
        ants.log_time("FINAL")
        
    def diffuse_food(self, ants):
        # setting up diffusion
        dif_setup_begin = time.time()
        
        scent_field = MyBot.U_DECAY * self.u_scent_field
        diff_factor = MyBot.DIFF_FACTOR
        
        # clamped source setup
        clamps = []
        clamp_field = np.empty(ants.dimensions)
        for ant in self.normal_ants:
            clamp_field[ant] = -1.0
        clamps.extend(self.normal_ants)
        for food in ants.food_set:
            clamp_field[food] = 50.0
        clamps.extend(ants.food_set)
        #for hill_loc in ants.enemy_hills():
            #clamp_field[hill_loc] = 100.0
        #clamps.extend(ants.enemy_hills())
        clamp_index = zip(*clamps)
                
        # unblocked and diff fields
        unblocked_field = ants.passable_field # WATER blocks diffusion
        scent_field[~unblocked_field] = 0.0 # reset scent of water (unnecessary)
        
        unblocked_n = (np.roll(unblocked_field, 1, axis=0) & 
                       unblocked_field)
        unblocked_s = (np.roll(unblocked_field,-1, axis=0) &
                       unblocked_field)
        unblocked_w = (np.roll(unblocked_field, 1, axis=1) &
                       unblocked_field)
        unblocked_e = (np.roll(unblocked_field,-1, axis=1) &
                       unblocked_field)
        adj_field = (unblocked_n.astype(np.int8) + unblocked_s + 
                       unblocked_w + unblocked_e)
        diff_loss = 1 - (diff_factor * adj_field)
        diff_n = diff_factor * unblocked_n
        diff_s = diff_factor * unblocked_s
        diff_w = diff_factor * unblocked_w
        diff_e = diff_factor * unblocked_e

        dif_setup_time = 1000*(time.time()-dif_setup_begin)
        log.info("Diff setup: %s ms", dif_setup_time)
        
        dif_begin = time.time()
        for diffusion in xrange(MyBot.U_ITERATIONS):
            # source blocks
            scent_field[clamp_index] = clamp_field[clamp_index]
            # build new scents
            n = np.concatenate((scent_field[-1:], scent_field[:-1]), axis=0)
            s = np.concatenate((scent_field[1:], scent_field[:1]), axis=0)
            w = np.concatenate((scent_field[:,-1:], scent_field[:,:-1]), axis=1)
            e = np.concatenate((scent_field[:,1:], scent_field[:,:1]), axis=1)
            scent_field = (diff_loss * scent_field
                + (diff_n * n)
                + (diff_s * s)
                + (diff_w * w)
                + (diff_e * e))
        self.u_scent_field = scent_field # store scent field
        dif_time = 1000*(time.time()-dif_begin)
        log.info("Unified: %s diff iterations: %s ms", MyBot.U_ITERATIONS, dif_time)
        
    def diffuse_explore(self, ants):
        # setting up variables
        dif_setup_begin = time.time()
        
        diff_factor = MyBot.E_FACTOR
        invisible = ~ants.visible_field
        unexplored = ~ants.explored_field
        # clamped source setup
        clamp_index = unexplored.copy()
        for e_ant in ants.enemy_ants():
            clamp_index[e_ant] = True
        clamp_field = 3.0 * clamp_index
        clamp_useless = (np.roll(clamp_index, 1, axis=0) &
                         np.roll(clamp_index,-1, axis=0) &
                         np.roll(clamp_index, 1, axis=1) &
                         np.roll(clamp_index,-1, axis=1))
        clamp_index = clamp_index - clamp_useless
        
        # re-initialize scent field with 1.0 at invisible squares
        scent_field = invisible.astype(np.float64) 
        
        # preset unblocked and diff fields
        # WATER or ant : do not receive diffusion
        unblocked_field = ants.passable_field & (~ ants.ant_field)
        scent_field[~unblocked_field] = 0.0 # reset ant/water scent
        
        unblocked_n = (np.roll(unblocked_field, 1, axis=0) & 
                       unblocked_field)
        unblocked_s = (np.roll(unblocked_field,-1, axis=0) &
                       unblocked_field)
        unblocked_w = (np.roll(unblocked_field, 1, axis=1) &
                       unblocked_field)
        unblocked_e = (np.roll(unblocked_field,-1, axis=1) &
                       unblocked_field)
        adj_field = (unblocked_n.astype(np.int8) + unblocked_s + 
                       unblocked_w + unblocked_e)
        diff_loss = 1 - (diff_factor * adj_field)
        diff_n = diff_factor * unblocked_n
        diff_s = diff_factor * unblocked_s
        diff_w = diff_factor * unblocked_w
        diff_e = diff_factor * unblocked_e
        
        dif_setup_time = 1000*(time.time()-dif_setup_begin)
        log.info("Diff setup: %s ms", dif_setup_time)
        
        dif_begin = time.time()
        log.debug(len(clamp_index.nonzero()[0]))
        for diffusion in xrange(MyBot.E_ITERATIONS):
            # source blocks
            scent_field[clamp_index] = 3.0 # should get from clamp_field
            # build new scents
            n = np.concatenate((scent_field[-1:], scent_field[:-1]), axis=0)
            s = np.concatenate((scent_field[1:], scent_field[:1]), axis=0)
            w = np.concatenate((scent_field[:,-1:], scent_field[:,:-1]), axis=1)
            e = np.concatenate((scent_field[:,1:], scent_field[:,:1]), axis=1)
            scent_field = (
                (diff_loss * scent_field)
                + (diff_n * n)
                + (diff_s * s)
                + (diff_w * w)
                + (diff_e * e))
        self.e_scent_field = scent_field # store scent field
        dif_time = 1000*(time.time()-dif_begin)
        log.info("Explore: %s diff iterations: %s ms", MyBot.E_ITERATIONS, dif_time)

    def diffuse_attack(self, ants):
        dif_setup_begin = time.time()
        
        my_hills = ants.my_hills()
        target_hills = ants.enemy_hills()
        enemies = ants.enemy_ants()
        
        if not self.enemy_seen:
            if len(target_hills) == 0 and len(enemies) == 0:
                return
            else:
                self.enemy_seen = True
        
        if my_hills:
            hilldist_begin = time.time()
            self.hilldist_calc(ants, limit=15)
            hilldist_time = 1000*(time.time()-hilldist_begin)
            log.debug("hilldist generation time: %s ms", hilldist_time)
            threat_value = {0:0.0, 1:90.0, 2:85.0, 3:80.0, 4:75.0, 
                            5:70.0, 6:64.0, 7:60.0, 8:50.0, 9:48.0, 
                            10:46.0, 11:44.0, 12:42.0, 13:40.0, 14:38.0,
                            13:36.0, 14:34.0, 15:32.0, INF:25.0}
            for dist in threat_value.keys():
                # multiple hills -> less value
                threat_value[dist] /= len(my_hills)
        
        if enemies:
            combat_safety_begin = time.time()
            self.gen_combat_safety(ants)
            combat_safety_time = 1000*(time.time()-combat_safety_begin)
            log.debug("Combat Safety generation: %s ms", combat_safety_time)
                
        # scaling setup
        coop_factor = MyBot.COOP_FACTOR
        coop_index = zip(*ants.my_ants())
                
        # unblocked and diff fields
        diff_factor = MyBot.A_FACTOR
        unblocked_field = ants.passable_field # WATER blocks diffusion
        unblocked_n = (np.roll(unblocked_field, 1, axis=0) & 
                       unblocked_field)
        unblocked_s = (np.roll(unblocked_field,-1, axis=0) &
                       unblocked_field)
        unblocked_w = (np.roll(unblocked_field, 1, axis=1) &
                       unblocked_field)
        unblocked_e = (np.roll(unblocked_field,-1, axis=1) &
                       unblocked_field)
        adj_field = (unblocked_n.astype(np.int8) + unblocked_s + 
                       unblocked_w + unblocked_e)
        diff_loss = 1 - (diff_factor * adj_field)
        diff_n = diff_factor * unblocked_n
        diff_s = diff_factor * unblocked_s
        diff_w = diff_factor * unblocked_w
        diff_e = diff_factor * unblocked_e

        # clamped source setup
        clamps = []
        clamp_field = np.empty(ants.dimensions)
        for hill_loc in target_hills:
            clamp_field[hill_loc] = 50.0
        clamps.extend(target_hills)
        

        # scent field initialize
        scent_field = MyBot.A_DECAY * self.a_scent_field
        scent_field[~unblocked_field] = 0.0 # reset scent of water (unnecessary)
        if my_hills:
            for ant in enemies:
                # affected by distance from my_hills
                if ant.hilldist < INF:
                    clamp_field[ant] = threat_value[ant.hilldist]
                    clamps.append(ant)
                else:
                    scent_field[ant] += threat_value[ant.hilldist]
        clamp_index = zip(*clamps)

        dif_setup_time = 1000*(time.time()-dif_setup_begin)
        log.info("Diff setup: %s ms", dif_setup_time)
        
        # iteration
        dif_begin = time.time()
        for diffusion in xrange(MyBot.A_ITERATIONS):
            # source blocks
            scent_field[clamp_index] = clamp_field[clamp_index]
            # build new scents
            n = np.concatenate((scent_field[-1:], scent_field[:-1]), axis=0)
            s = np.concatenate((scent_field[1:], scent_field[:1]), axis=0)
            w = np.concatenate((scent_field[:,-1:], scent_field[:,:-1]), axis=1)
            e = np.concatenate((scent_field[:,1:], scent_field[:,:1]), axis=1)
            scent_field = (diff_loss * scent_field
                + (diff_n * n)
                + (diff_s * s)
                + (diff_w * w)
                + (diff_e * e))
            # scaling
            scent_field[coop_index] *= coop_factor
        dif_time = 1000*(time.time()-dif_begin)
        log.info("Attack: %s diff iterations: %s ms", MyBot.A_ITERATIONS, dif_time)
        
        # store scent field
        self.a_scent_field = scent_field

    def hilldist_calc(self, ants, limit=10):
        cur_dist = 0
        my_hills = ants.my_hills()
        # hill_num = len(my_hills)
        queue = list(my_hills)
        complete = set()
        for cell in ants.world_list:
            cell.hilldist = INF
        while queue and cur_dist <= limit:
            new_queue = []
            while queue:
                cell = queue.pop()
                if cell not in complete:
                    cell.hilldist = cur_dist
                    new_queue.extend(cell.adj)
                    complete.add(cell)
            queue = new_queue
            cur_dist += 1
        
    def gen_combat_safety(self, ants):
        player_ants = []
        positions = []
        for owner in xrange(10):
            player_ants.append([a for a in ants.ant_set if a.contents == owner])
        
        attack_stamp = ants.attack_stamp
        combat_stamp = self.approx_combat_stamp
        attack_field = self.attack_field
        weakness_field = self.weakness_field
        best_enemy = self.lowest_enemy_weakness
        safety_field = self.safety_field
        attack_field.fill(0)
        weakness_field.fill(0)
        best_enemy.fill(INF)
        safety_field.fill(SAFE)
        
        for owner in xrange(10):
            player_attack = attack_field[owner]
            for ant in player_ants[owner]:
                increment_stamp(player_attack, combat_stamp, ant)
        for owner in xrange(10):
            weakness_field[owner] = sum(attack_field[p] 
                                    for p in xrange(10) if p != owner)
            arena = set(loc for a in player_ants[owner]
                                for loc in a.moves
                                    if weakness_field[owner][loc] > 0)
            positions.append(arena)
        for enemy in xrange(1,10):
            for enemy_position in positions[enemy]:
                min_stamp(best_enemy, 
                          (attack_stamp, weakness_field[enemy][enemy_position]),
                          enemy_position)
        my_weakness = weakness_field[ME]
        safety_field[my_weakness > best_enemy] = DIE
        safety_field[my_weakness == best_enemy] = NEUTRAL
        
        self.soldiers = set(a for a in player_ants[ME] 
                                  if positions[ME].intersection(a.moves))

if __name__ == '__main__':
    # psyco will speed up python a little, but is not needed
    try:
        import psyco
        psyco.full()
    except ImportError:
        pass
    
    init_options()
    
    try:
        # if run is passed a class with a do_turn method, it will do the work
        # this is not needed, in which case you will need to write your own
        # parsing function and your own game state class
        Ants.run(MyBot())
    except KeyboardInterrupt:
        print('ctrl-c, leaving ...')
